metaflow/plugins/resources_decorator.py

```diff
@@ -0,0 +1,62 @@
+from metaflow.decorators import StepDecorator
+
+
+class ResourcesDecorator(StepDecorator):
+    """
+    Step decorator to specify the resources needed when executing this step.
+    This decorator passes this information along when requesting resources
+    to execute this step.
+
+    This decorator can be used for two purpose:
+    When using with AWS Batch decorator, only 'cpu', 'gpu', and 'memory' parameters are supported.
+    When using for Kubeflow Pipeline, 'cpu', and 'memory' sets resource requests;
+        'cpu_limit', 'gpu', 'memory_limit' sets resource limit.
+        For more details please refer to
+        https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+    This decorator is ignored if the execution of the step does not happen on Batch.
+    To use, annotate your step as follows:
+    ```
+    @resources(cpu=32, memory="15G")
+    @step
+    def myStep(self):
+        ...
+    ```
+
+    Parameters
+    ----------
+    cpu : Union[int, float, str]
+        AWS Batch: Number of CPUs required for this step. Defaults to 1. Must be integer.
+        KFP: Number of CPUs required for this step. Defaults to 1. Can input int, float, or str.
+            Support millicpu requests using float or string ending in 'm'.
+            A request with a decimal point, like 0.1, is converted to 100m by the API,
+            and precision finer than 1m is not allowed.
+    cpu_limit : Union[int, float, str]
+        Not for AWS Batch.
+        KFP: Number of CPUs limited for this step. Defaults None - relying on Kubernetes defaults.
+    gpu : int
+        AWS Batch: Number of GPUs required for this step. Defaults to 0.
+        KFP: GPU limit for this step. Defaults to 0.
+            GPU are only supposed to be specified in limit section.
+            See https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/
+    gpu_vendor : str
+        Not for AWS Batch.
+        KFP: "nvidia" or "amd". Defaults to "nvidia".
+    memory : Union[int, str]
+        AWS Batch: Memory size (in MB) required for this step. Defaults to 4000.
+        KFP: Memory size required for this step. Default to 4000 MB. For int input, unit defaults to MB.
+            More memory units are supported, including "E", "P", "T", "G", "M", "K". (i.e. "4000M")
+    memory_limit : Union[int, str]
+        Not for AWS Batch.
+        KFP: Memory limit for this step. Default unit is MB.
+            More memory units are supported, including "E", "P", "T", "G", "M", "K". (i.e. "4000M")
+             Defaults None - relying on Kubernetes defaults.
+    """
+    name = 'resources'
+    defaults = {
+        "cpu": "1",
+        "cpu_limit": None,
+        "gpu": "0",
+        "gpu_vendor": "nvidia",
+        "memory": "4000",
+        "memory_limit": None,
+    }
```

Comment:
What would be a good set of default values? \
Setting through Metaflow is easier for model developers, since they wouldn't worry too much about Kubernetes stuff. \
Alternatively giving a `None` value here makes it dependent on defaults from KFP or kubernetes namespaces, making defaults value easier to manage for AIP teams.

Reply:
These look pretty good to me. I think relying on `None` to defer to the platform defaults seems the most sane to me.

Reply:
Actually maybe should we also defer the `gpu` value too, seems unlikely but potentially if we set it to `0` by default we could stop a pod scheduling on a default environment with a `GPU`. Seems unlikely to me but I'm happy either way :) 

Mostly it'd allow us to remove this special case: https://github.com/zillow/metaflow/pull/20/files#diff-8fe7ca325ce1b6d4fe735b509f0bc7b7R154

Reply:
Good catch for GPU count. "gpu=0" was the default for AWS batch plug in, so it was kept. In KFP sdk it's not possible to set GPU limit to zero, though by default no GPU would be assigned: https://github.com/kubeflow/pipelines/blob/master/sdk/python/kfp/dsl/_container_op.py#L319

Kubernete also has this page talking about GPU scheduling: https://kubernetes.io/docs/tasks/manage-gpus/scheduling-gpus/, mentioning that "Containers (and Pods) do not share GPUs. There's no overcommitting of GPUs." Maybe setting default GPU count to positive number is not a good idea.

Reply:
Removed defaults completely from `@resource` decorator. Each platform using this resource decorators have to implement defaults individually. No extra defaults were set for KFP so that the environment config can be used.

AWS batch doesn't seems affected: the same default was written in`@resource` decorator and `@batch` decorator. Defaults from `@batch` will keep functionalities unchanged.